cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_heap_dpc.cpp

```diff
@@ -0,0 +1,448 @@
+                       dst_t* const dst_ptr,
+                       std::int32_t k,
+                       std::int32_t width,
+                       std::int32_t height,
+                       std::int32_t src_str,
+                       std::int32_t ids_str,
+                       std::int32_t dst_str,
+                       acc_t heaps)
+            : base_t(ids_str, dst_str, ids_ptr, dst_ptr),
+              src_str_(src_str),
+              src_ptr_(src_ptr),
+              k_(k),
+              width_(width),
+              height_(height),
+              heaps_(heaps) {}
+
+    void operator()(sycl::nd_item<1> item) const {
+        using sycl::ONEAPI::reduce;
+
+        auto sg = item.get_sub_group();
+
+        const std::int32_t sid = sg.get_group_linear_id();
+        const std::int32_t cid = sg.get_local_linear_id();
+        const std::int32_t sg_width = sg.get_local_range().size();
+        const std::int32_t rid = sid + item.get_group_linear_id() * sg.get_group_range().size();
+
+        if ((rid >= height_) || (cid >= sg_width))
+            return;
+
+        dst_t pbuff_dst[pbuff_size] = { dst_default };
+        idx_t pbuff_ids[pbuff_size] = { idx_default };
+        std::int32_t pbuff_count, prev_count;
+
+        sel_t* const heaps = heaps_.get_pointer();
+        sel_t* const curr_heap = heaps + (k_ + 1) * sid;
+        sel_t* const curr_heap_end = curr_heap + (k_ + 1);
+
+        // Heap initialization
+        for (std::int32_t i = cid; i <= k_; i += sg_width) {
+            auto& values = curr_heap[i];
+            values.idx = idx_default;
+            values.dst = dst_default;
+        }
+        sg.barrier();
+
+        std::int32_t k_written = 0;
+        dst_t worst_val = dst_default;
+        const auto step = sg_width * pbuff_size;
+        const auto* const row = src_ptr_ + rid * src_str_;
+        for (std::int32_t i = 0; i < width_; i += step) {
+            const std::int32_t block_start_col = cid + i;
+
+            pbuff_count = 0, prev_count = 0;
+            worst_val = curr_heap->dst;
+
+            // Collecting temporary best values in private memory
+            for (std::int32_t j = 0; (j < pbuff_size); ++j) {
+                const idx_t idx = block_start_col + sg_width * j;
+                const bool handle = idx < width_;
+                const dst_t val = handle ? *(row + idx) : dst_default;
+                pbuff_count += bool(val < worst_val);
+                pbuff_ids[prev_count] = idx;
+                pbuff_dst[prev_count] = val;
+                prev_count = pbuff_count;
+            }
+            sg.barrier();
+
+            // writing temporary best values into heap
+            std::int32_t cid_to_handle, handle_this;
+            do {
+                handle_this = pbuff_count ? cid : -1;
+                k_written = reduce(sg, k_written, max_func);
+                cid_to_handle = reduce(sg, handle_this, max_func);
+                if (cid_to_handle == cid) {
```

Comment:
Processing on a single work-item?

Reply:
1) It works only if the WI accumulates enough suitable points
2) There is no way to do it synchronously. Heap can operate only in single thread